icu4c/source/common/rbbi_cache.cpp

```diff
@@ -420,50 +426,141 @@ UBool RuleBasedBreakIterator::BreakCache::populateNear(int32_t position, UErrorC
     return true;
 }
 
+static inline bool isNumber(UChar32 ch) {
+    UErrorCode status = U_ZERO_ERROR;
+    static UnicodeSet numberSet;
+    if(numberSet.isEmpty()) {
+        numberSet.applyPattern(UnicodeString(u"[:Nd:]"), status);
+        if (U_SUCCESS(status)) {
+            numberSet.compact();
+        }
+    }
+    return numberSet.contains(ch);
+}
+
+static inline bool isCj(UChar32 ch) {
+    UErrorCode status = U_ZERO_ERROR;
+    static UnicodeSet cjSet;
+    if(cjSet.isEmpty()) {
+        cjSet.applyPattern(UnicodeString(u"[[:Han:][:Hiragana:][:Katakana:]\u30fc\uff70\uff9e\uff9f]"), status);
+        if (U_SUCCESS(status)) {
+            cjSet.compact();
+        }
+    }
+    return cjSet.contains(ch);
+}
+
+/* getting a uint32_t properties word from the data */
+#define GET_PROPS(c, result) ((result)=UTRIE2_GET16(&propsTrie, c))
+
+/** Mask constant for multiple UCharCategory bits (OP Open Punctuation). @stable ICU 2.1 */
+#define U_GC_OP_MASK \
+            (U_GC_PI_MASK|U_GC_PS_MASK)
+#define U_GC_CP_MASK \
+            (U_GC_PD_MASK|U_GC_PE_MASK|U_GC_PC_MASK|U_GC_PO_MASK|U_GC_PF_MASK)
+
+static UBool u_isOpenPunct(UChar32 c) {
+    uint32_t props;
+    GET_PROPS(c, props);
+    return (UBool)((CAT_MASK(props)&U_GC_OP_MASK)!=0);
+}
+
+static UBool u_isClosePunct(UChar32 c) {
+    uint32_t props;
+    GET_PROPS(c, props);
+    return (UBool)((CAT_MASK(props)&U_GC_CP_MASK)!=0);
+}
+
+UBool RuleBasedBreakIterator::BreakCache::handleNumber(int32_t pos) {
```

Comment:
I would like to figure out some way of moving logic like this, which is specific to one type of break iterator, out of this level (BreakCache), which is completely general to all break iterators. If we can figure out how, it would be better appearing in the Break Engine and/or the rules for its type of break iterator.

I have some partial thoughts, but they aren't worked out. I'll see if I can come up with something.

Reply:
Move the logic to CjkBreakEngine.